src/shogun/mathematics/graph/Tensor.cpp

```diff
@@ -0,0 +1,15 @@
+#include <shogun/mathematics/graph/Tensor.h>
+
+#include <shogun/io/fmt/fmt.h>
+
+using namespace shogun;
+
+std::string Tensor::to_string() const
+{
+	return fmt::format("Tensor(shape={}, type={})", m_shape, m_type);
+}
+
+template Tensor::Tensor(const SGVector<float32_t>&);
```

Comment:
since the only function atm of Tensor in the implementation part is to_string don't you think that header only Tensor (all the implementation is there) would be better?

Reply:
this was so that fmt wouldn't become a runtime dependency. But I guess Tensor is used in some other .cpp files

Reply:
mmm ok i see... lets keep it as is and then we'll refactor lets just push this to a level that things are functional and then we cleanup on the end imo